src/paths/floyd_warshall.c

```diff
@@ -140,3 +141,160 @@ igraph_error_t igraph_distances_floyd_warshall(
+            if (mode == IGRAPH_ALL) {
+                IGRAPH_ERRORF("Negative edge weight (%g) found in undirected graph "
+                              "while calculating distances with Floyd-Warshall.",
+                              IGRAPH_ENEGLOOP, w);
+            } else if (to == from) {
+                IGRAPH_ERRORF("Self-loop with negative weight (%g) found "
+                              "while calculating distances with Floyd-Warshall.",
+                              IGRAPH_ENEGLOOP, w);
+            }
+        }
+
+        if (out && MATRIX(*res, from, to) > w) MATRIX(*res, from, to) = w;
+        if (in && MATRIX(*res, to, from) > w) MATRIX(*res, to, from) = w;
+    }
+
+    /* predecessors[u][v] is the second but last vertex on the shortest path from u to v */
+    igraph_matrix_int_t predecessors;
+    IGRAPH_MATRIX_INT_INIT_FINALLY(&predecessors, no_of_nodes, no_of_nodes);
+
+    /* children[u][i] is the i-th children of u in a tree of shortest distances
+       rooted at k in the main loop below */
+    igraph_matrix_int_t children;
+    IGRAPH_MATRIX_INT_INIT_FINALLY(&children, no_of_nodes, no_of_nodes);
+
+    /* no_of_children[u] is the number of children of u
+       rooted at k in the main loop below */
+    igraph_vector_int_t no_of_children;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&no_of_children, no_of_nodes);
+
+    igraph_vector_int_t dfs_traversal;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&dfs_traversal, no_of_nodes);
+
+    igraph_vector_int_t dfs_skip;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&dfs_skip, no_of_nodes);
+
+    igraph_stack_int_t stack;
+    IGRAPH_STACK_INT_INIT_FINALLY(&stack, no_of_nodes);
+
+    for (igraph_integer_t v=0; v < no_of_nodes; v++) {
+        for (igraph_integer_t u=0; u < no_of_nodes; u++) {
+            MATRIX(predecessors, u, v) = u;
+        }
+    }
+
+    for (igraph_integer_t k=0; k < no_of_nodes; k++) {
+        /* resetting no_of_children vector */
+        for (igraph_integer_t parent=0; parent < no_of_nodes; parent++) {
+            VECTOR(no_of_children)[parent] = 0;
+        }
+        /* constructing the tree out_k (as in the paper) but
+           representing it as the children matrix */
+        for (igraph_integer_t v=0; v < no_of_nodes; v++) {
+            if (k == v) continue;
+            igraph_integer_t parent = MATRIX(predecessors, k, v);
+            MATRIX(children, parent, VECTOR(no_of_children)[parent]) = v;
+            VECTOR(no_of_children)[parent]++;
+        }
+        /* constructing dfs-traversal and dfs-skip arrays
+            for the out_k tree
+        */
+        IGRAPH_CHECK(igraph_stack_int_push(&stack, k));
+        igraph_integer_t counter = 0;
+        while (!igraph_stack_int_empty(&stack)) {
+            igraph_integer_t parent = igraph_stack_int_pop(&stack);
+            if (parent > -1) {
+                VECTOR(dfs_traversal)[counter] = parent;
+                counter++;
+                IGRAPH_CHECK(igraph_stack_int_push(&stack, -parent - 1));
```

Comment:
Why is the pushed value negated here?

Reply:
We use `stack` to construct both `dfs_traversal` and `dfs_skip` array. The non-negative values are used to construct `dfs_traversal` array, and the negative values `dfs_skip` array. In particular, values `-parent - 1` can be thought of as markers that are popped from the stack right after all the descendants of `parent` were processed. At that point `counter` stores the number of vertices seen up to that point. Therefore `VECTOR(dfs_skip)[-(parent + 1)] = counter` sets the index of the first vertex in `dfs_traversal` following `parent` that is not a descendant of the `parent` as desired.

Reply:
I can put there a comment explaining that.

Reply:
Can you please add a short comment in the code that explains this?